### ISSUE TITLE
New syntax for sealed hierarchies which span files

### ISSUE THREAD

**Issue by jchayat** [typelevel/scala#76]
Currently, the Scala compiler requires all children of a sealed class or trait to be defined at the same file. This requirement may prevent breaking code into multiple files which may result in very long files when dealing with large hierarchies.

I propose an extended syntax for the sealed keyword to resolve this limitation.

``` scala
// A.scala
package foo.bar

sealed[foo.bar, F, G, H#HH, H.HHH] abstract class A

class B extends A // Legal - B is in the same file as A

// C.scala
package foo.bar

class C extends A // Legal - C is in the foo.bar package which is in the sealed scope

object C extends A // Legal - C is in the foo.bar package which is in the sealed scope

// D.scala
package foo.bar.baz

class D extends A // Legal - D is in a subpackage of the foo.bar package

// E.scala
package foo

class E extends A // Illegal - the foo package which is not in the sealed scope

// F.scala
package foo

class F extends A // Legal - F is in the sealed scope

// G.scala
package foo

class G { // G is in the sealed scope although it does not extend A
  object GG extends A // Legal - GG is an inner object of G which is in the sealed scope
}

object G { // G is in the sealed scope although it does not extend A
  trait GGG extends A // Legal - GGG is an inner trait of G which is in the sealed scope
}

// H.scala
package foo

class H {
  object HH extends A // Legal - H#HH is in the sealed scope
}

object H { // G is in the sealed scope although it does not extend A
  trait HHH extends A // Legal - H.HHH is in the sealed scope
}

// I.scala
package foo

sealed[foo.bar] abstract trait I

class J extends I // Legal - J in the same file as I

// K.scala
package foo

trait K extend I // Illegal - package foo is not in the sealed scope
```

**Comment by paulp** [comment]
For more anti-compositional restrictions to lift see https://issues.scala-lang.org/browse/SI-5962 and https://issues.scala-lang.org/browse/SI-7348 .

**Comment by jchayat** [comment]
@paulp I didn't understand the relationship to this proposal.

**Comment by paulp** [comment]
You can't break anything which contains a value class into multiple files because it won't be a value class upon composition since the compiler won't honor the enclosing self-type which includes the statically reachable provision. This is exactly analogous to the problem with sealed.

The second one is an aspect of the general phenomenon that companion objects are insanely privileged yet must be defined in the same file as the companion class, which leads to endless gyrations trying to reach those privileges without overspecification.

**Comment by jchayat** [comment]
I see

**Comment by cvogt** [comment]
-1 to syntactic support for that. I think either detect it automatically or keep the restriction. Also this should be not very common and even long files aren't the end of the world. Not enough reason to add syntax and thereby perceived complexity IMHO.

**Comment by jchayat** [comment]
@cvogt IMO, global auto detection is counter productive - the whole point of sealed is restricting who may subclass. In a way, allowing packages in sealed scope acts as auto detection in a limited and controlled scope.

**Comment by cvogt** [comment]
Auto detection shouldn't be global, but restricted to the currently compiled project. Anything else can't be included anyways. But I suspect this to be too complicated to implement to be feasible. (Restricting to a package could be done by making the sealed trait package private).

**Comment by aloiscochard** [comment]
I'm a bit concerned about how that would work with the macro enumerating
sealed objects.

There is some problems with it already as it depends on the order of
definition.

**Comment by jchayat** [comment]
@cvogt I want to achieve safety within a project, particularly in a large project with a lot of people working on it.
Visibility is an orthogonal concern - I _want_ my trait to be visible from the outside, that's the whole point.

**Comment by jchayat** [comment]
@aloiscochard AFAIK that macro is already broken. I would love to see it working, but I don't see why sealedness should be coupled to source files.

The compiler should look at all relevant files and check whether the rules are followed, then construct a proper sealed hierarchy.

**Comment by cvogt** [comment]
@jchayat You are right. Restricting visibility wouldn't make much sense.

**Comment by Blaisorblade** [comment]
TL;DR. I fear once you specify this in enough detail, the complexity might be more than it seems, in particular because of exhaustiveness checking, and I don't see yet a strong enough motivation for it. I don't doubt you'd like the feature, but I am not even sure how you plan to use it, even less how I should use it. For all use cases I understand, having one long file would work, and the example given is not an interesting use case. Sorry this post is so long.

> The compiler should look at all relevant files and check whether the rules are followed, then construct a proper sealed hierarchy.

Problems: 
1) That algorithm doesn't respect separate compilation. Fixing the checking part is simple: when defining a class C extending other types, you can check that the sealedness constraints of those parent types are satisfied.
2) What's a proper sealed hierarchy? It seems you needn't do any transformation, you can just preserve the sealed annotation in the output — however standard Scalac wouldn't understand the annotation, and I'm not sure what to do there. What are you proposing here?

> I don't see why sealedness should be coupled to source files.

Sealing a type T has two goals, and for one coupling with source files makes quite some sense.
- preventing other modules from declaring new subtypes of T. Here you're right.
- allow checking that clients of T cover all possible cases when pattern matching. This becomes harder if you don't couple sealing with source files: you need to know all subtypes of T before checking client code, and I'm not sure what becomes of this under your proposal and separate compilation. Specifically, let's say I compile T, then a client pattern matching on T, then a module defining new allowed subtypes of T. What should happen? I think "the compiler loads the sealed scope" might be a complex route, especially because avoiding cyclic dependencies isn't easy.

> @cvogt I want to achieve safety within a project, particularly in a large project with a lot of people working on it.

@jchayat Can you elaborate? What kind of safety? I'll ignore the "lots of people" issue and reason in terms of modularity, since the mapping between modules and people is not interesting to a Scala compiler.

You want to protect client module A so that module B can't inject subtypes, but module C can?

Why is that kind of restriction generally desirable? And why in that form? Here's an alternative, assuming 

In https://issues.scala-lang.org/browse/SI-5962, Paul mentions the strategy for how to break up an object into different traits. Implementing a `@sealedToContainingObject` would be easier: the actual sealed hiearchy lives in the completed object, and all the problems I mentioned disappear in a puff. (As long as you only try to pattern match on the member of the complete object).

Demo code follows.

``` scala
trait Base {
  this: ScalaLanguage.type =>
  @sealedToContainingObject trait AST
}

trait Expressions {
  this: ScalaLanguage.type =>
  case object NullLiteral extends AST
  //...
}

trait Statements {
  this: ScalaLanguage.type =>
  //...
}

object ScalaLanguage extends Base with Expressions with Statements

//client code
import ScalaLanguage.AST

def f(tree: AST) =
  tree match { //this pattern match can be checked for exhaustiveness, because this code depends on ScalaLanguage, hence it sees the complete definition.
    //...
  }

def g(tree: Base#AST) = {
  //I'm not sure you should be able to match on tree here — certainly you can't expect exhaustiveness checking here.
  ...
}
```

The different top-level definitions can live in different files & packages, of course.

**Comment by milessabin** [comment]
To resurrect this issue, please rework it as an issue/PR against Lightbend Scala (ie. scala/scala).
